lib/spack/docs/configuration.rst

```diff
@@ -221,3 +269,145 @@ The syntax for the ``provider`` section differs slightly from other
+   the "factory" settings. Customizations should be made in the other
+   higher precedence scopes, as defaults may change from version to
+   version.
+
+2. **site**: Stored in ``$(prefix)/etc/spack/``.  Settings here affect
+   only *this instance* of Spack.  They override ``defaults``.  The site
+   scope can can be used for per-project settings (one spack instance per
+   project) or for site-wide settings on a multi-user machine (e.g., for
+   a common spack instance).
+
+3. **user**: Stored in the home directory, ``~/.spack/``. These settings
+   affect all instances of Spack and take the highest precedence.
+
+When configurations conflict, settings from higher-precedence scopes
+override lower-precedence settings.
+
+Commands that modify scopes (e.g., ``spack compilers``, ``spack repo``,
+etc.) take a ``--scope=<name>`` parameter that you can use to control
+which scope is modified.
+
+.. _platform-scopes:
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+Platform-specific scopes
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For each scope above, there can *also* be platform-specific settings. For
+example, on Blue Gene/Q machines, Spack needs to know the location of
+cross-compilers for the compute nodes.  This configuration is in
+``etc/spack/defaults/bgq/compilers.yaml``.  It will take precedence over
+settings in the ``defaults`` scope, but can still be overridden by
+settings in ``site``, ``site/bgq``, ``user``, or ``user/bgq``. So, the
+full scope precedence is:
+
+1. ``defaults``
+2. ``defaults/<platform>``
+3. ``site``
+4. ``site/<platform>``
+5. ``user``
+6. ``user/<platform>``
+
+Each configuration directory may contain several configuration files,
+such as ``config.yaml``, ``compilers.yaml``, or ``mirrors.yaml``.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Configuration file format
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Configuration files are written in YAML.  For more details on the format,
+see `yaml.org <http://yaml.org>`_ and `libyaml
+<http://pyyaml.org/wiki/LibYAML>`_.  Each file is structured as nested
+dictionaries and lists.  e.g., ``config.yaml`` might look like this:
+
+.. code-block:: yaml
+
+   config:
+     install_tree: $spack/opt/spack
+     module_roots:
+       lmod:   $spack/share/spack/lmod
+     build_stage:
+       - $tempdir
+       - /nfs/tmp2/$user
+
+Note that everything in the config file needs to be nested under a
+top-level section corresponding to the config file's name.  So,
+``config.yaml`` starts with ``config:``, and ``mirrors.yaml`` starts with
+``mirrors:``.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Merging scopes
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, Spack attempts to merge configurations across scopes.  That
+is, if there are ``config.yaml`` files in both the default and site
+scopes, Spack will merge the settings between the two files.  You can see
+the results of this merge with the ``spack config get <configtype>``
+command. For example, if your configurations look like this:
+
```

Comment:
Most of this stuff has been there -- it just wasn't documented outside `config.py`.  I realized that it probably needed to go in the real docs, since the config system in Spack is actually halfway decent at taking options from many different places/roles and combining them.